### PR TITLE
EDSC-4530: Move admin retrievals to graphql/apollo

### PR DESCRIPTION
**Pull request by trevorlang**
# Overview

### What is the feature?

Move admin retrievals list and details page to graphql/apollo

### What is the Solution?

Move admin retrievals list and details page to graphql/apollo

### What areas of the application does this impact?

Admin retrievals list and retrieval details

# Testing

### Reproduction steps

- **Environment for testing:** Any
- **Collection to test with:** n/a

1. Confirm admin retrievals list displays retrievals
2. Confirm admin retrievals filters filter the list appropriately
3. Confirm the retrieval is loaded when clicked in the list
4. Confirm the retrieval displays its information properly, including retrieval collections and retrieval orders

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: serverless/src/adminGetRetrieval/__tests__/handler.test.js
Comment (trevorlang) [inline serverless/src/adminGetRetrieval/__tests__/handler.test.js:1]: Delete cdk resources for removed handlers

File: serverless/src/graphQl/resolvers/admin.js [hunk 1]
```diff
@@ -9,6 +43,112 @@ export default {
...
+      const { loaders } = context
+
+      const loaderData = await loaders.users.load(parent.userId)
+
+      return camelcaseKeys(loaderData, { deep: true })
+    },
+    obfuscatedId: async (parent) => {
+      const { id } = parent
+
+      return obfuscateId(id)
```
Comment (trevorlang) [inline serverless/src/graphQl/resolvers/admin.js:91]: Double check if this needs a seed value

File: serverless/src/graphQl/types/admin.graphql [hunk 2]
```diff
@@ -1,38 +1,262 @@
...
+  "Timestamp of when the retrieval was created"
+  createdAt: DateTime
+}
+
+input AdminRetrievalInput {
+  "Obfuscated ID to filter by"
+  obfuscatedId: String
+}
+
+type PageInfo {
```
Comment (trevorlang) [inline serverless/src/graphQl/types/admin.graphql]: This could be moved

File: serverless/src/graphQl/utils/databaseClient.js [hunk 3]
```diff
@@ -66,4 +128,97 @@ export default class DatabaseClient {
...
+      }
+
+      if (ursId) {
+        query = query.where({ 'users.urs_id': ursId })
+      }
+
+      if (retrievalCollectionId) {
+        query = query
+          .leftJoin('retrieval_collections', { 'retrievals.id': 'retrieval_collections.retrieval_id' })
+          .where({ 'retrieval_collections.id': parseInt(retrievalCollectionId, 10) })
```
Comment (trevorlang) [inline serverless/src/graphQl/utils/databaseClient.js]: This parseInt can probably be removed

File: serverless/src/graphQl/utils/getContext.js [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
 
 const { env } = getApplicationConfig()
 
 const databaseClient = new DatabaseClient()
 
+const loaders = getLoaders({ databaseClient })
+
+/**
+ * Gets the context for the GraphQL resolver
+ * @param {*} params The parameters for the context
```
Comment (trevorlang) [inline serverless/src/graphQl/utils/getContext.js]: Make this param better

File: serverless/src/graphQl/utils/getLoaders.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import DataLoader from 'dataloader'
+
+/**
+ * Gets the DataLoaders for the GraphQL resolvers
+ * @param {*} params The parameters for the DataLoaders
```
Comment (trevorlang) [inline serverless/src/graphQl/utils/getLoaders.js]: Make this param better

File: serverless/src/graphQl/utils/getLoaders.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * @returns The DataLoaders object
+ */
+const getLoaders = ({ databaseClient }) => ({
+  retrievalCollections: new DataLoader(async (ids) => {
```
Comment (trevorlang) [inline serverless/src/graphQl/utils/getLoaders.js:17]: Add comments for each loader

File: static/src/js/actions/admin/__tests__/retrievals.test.js
Comment (trevorlang) [inline static/src/js/actions/admin/__tests__/retrievals.test.js:1]: Add these back in if requeueOrder

File: static/src/js/actions/admin/retrievals.js [hunk 7]
```diff
@@ -1,178 +1,12 @@
-import RetrievalRequest from '../../util/request/admin/retrievalRequest'
-import { addToast } from '../../util/addToast'
-
-import {
-  SET_ADMIN_RETRIEVAL,
```
Comment (trevorlang) [inline static/src/js/actions/admin/retrievals.js:5]: Make sure these get removed from the constants

File: static/src/js/components/AdminRetrievalDetails/AdminRetrievalDetails.jsx [hunk 8]
```diff
@@ -38,177 +48,172 @@ export const AdminRetrievalDetails = ({
...
+                </Col>
+                <Col className="admin-retrieval-details__metadata-display-item">
+                  <span className="admin-retrieval-details__metadata-display-heading">Obfuscated ID</span>
+                  <span data-testid="admin-retrieval-details__metadata-display-content" className="admin-retrieval-details__metadata-display-content">{obfuscatedId}</span>
+                </Col>
+              </Row>
+              <Row className="admin-retrieval-details__metadata-display mb-2">
+                <Col className="admin-retrieval-details__metadata-display-item">
+                  <span className="admin-retrieval-details__metadata-display-heading">Source Path</span>
+                  <a data-testid="admin-retrieval-details__metadata-display-content" className="admin-retrieval-details__metadata-display-content" href={sourcePath} target="_blank" rel="noopener noreferrer">
```
Comment (trevorlang) [inline static/src/js/components/AdminRetrievalDetails/AdminRetrievalDetails.jsx]: Double check all test ids are removed

File: static/src/js/components/AdminRetrievals/AdminRetrievals.jsx [hunk 9]
```diff
@@ -30,38 +24,18 @@ export const AdminRetrievals = ({
       ]
     }
   >
-    <Row className="justify-content-end mb-2">
```
Comment (trevorlang) [inline static/src/js/components/AdminRetrievals/AdminRetrievals.jsx:33]: See what it looks like if AdminRetrievalsForm gets moved here

File: static/src/js/components/AdminRetrievals/AdminRetrievalsList.jsx [hunk 10]
```diff
@@ -1,183 +1,263 @@
...
+  const onUrsIdSortClick = useCallback(() => {
+    const sortKeyString = sortKey || ''
+
+    if (sortKeyString.indexOf('urs_id') < 0) {
+      setSortKey('-urs_id')
     }
 
-    if (sortKey === '+username') {
-      handleSort('')
+    if (sortKeyString === '+urs_id') {
```
Comment (trevorlang) [inline static/src/js/components/AdminRetrievals/AdminRetrievalsList.jsx]: Reorder these if statements to follow the order when clicking

File: static/src/js/containers/AdminRetrievalContainer/AdminRetrievalContainer.jsx [hunk 11]
```diff
@@ -1,62 +1,39 @@
...
-    const { id } = params
-
-    onFetchAdminRetrieval(id)
-  }, [])
-
   const { params } = match
-  const { id } = params
-
-  const { [id]: selectedRetrieval } = retrievals
+  const { id: obfuscatedId } = params
```
Comment (trevorlang) [inline static/src/js/containers/AdminRetrievalContainer/AdminRetrievalContainer.jsx]: Rename this in the params

File: serverless/src/graphQl/utils/databaseClient.js [hunk 12]
```diff
@@ -29,6 +31,159 @@ export default class DatabaseClient {
...
+    }
+  }
+
+  /**
+   * Retrieves retrievals based on the provided filters
+   * @param {Object} params The filters to apply
+   * @returns {Promise<Array>} A promise that resolves to an array of retrieval objects
+   */
+  async getRetrievals({
+    ursId,
```
Comment (macrouch) [inline serverless/src/graphQl/utils/databaseClient.js]: sort keys

File: serverless/src/graphQl/utils/getLoaders.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
+import DataLoader from 'dataloader'
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (macrouch) [inline serverless/src/graphQl/utils/getLoaders.js:2]: This is only being used as a type in a jsdoc, is it possible to do that in a way where we don't ignore the eslint rule?

File: static/src/js/components/AdminRetrievalDetails/AdminRetrievalDetails.jsx [hunk 14]
```diff
@@ -1,28 +1,38 @@
...
 import Row from 'react-bootstrap/Row'
 import Table from 'react-bootstrap/Table'
 
+import { gql, useQuery } from '@apollo/client'
 import { commafy } from '../../util/commafy'
 
 import Button from '../Button/Button'
 
 import './AdminRetrievalDetails.scss'
+import ADMIN_RETRIEVAL from '../../operations/queries/adminRetrieval'
```
Comment (macrouch) [inline static/src/js/components/AdminRetrievalDetails/AdminRetrievalDetails.jsx]: Import before scss

File: static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx [hunk 15]
```diff
@@ -1,180 +1,162 @@
...
-              order_number: '40058',
-              state: 'creating',
-              type: 'ECHO ORDERS'
-            }]
-          }]
+      const onRequeueOrderMock = jest.fn()
+
+      setup({
+        overrideProps: {
+          onRequeueOrder: onRequeueOrderMock
```
Comment (macrouch) [inline static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx]: You don't need to set this as a new mock function, `props.onRequeueOrder` is already a mock function from the defaultProps

File: static/src/js/components/AdminRetrievals/AdminRetrievalsList.jsx [hunk 16]
```diff
@@ -1,183 +1,265 @@
...
+  Form,
+  InputGroup
+} from 'react-bootstrap'
+import ADMIN_RETRIEVALS from '../../operations/queries/adminRetrievals'
 
 import EDSCIcon from '../EDSCIcon/EDSCIcon'
 
 import 'rc-pagination/assets/index.css'
 import './AdminRetrievalsList.scss'
+import Spinner from '../Spinner/Spinner'
```
Comment (macrouch) [inline static/src/js/components/AdminRetrievals/AdminRetrievalsList.jsx]: Import before scss

File: serverless/src/graphQl/resolvers/admin.js [hunk 17]
```diff
@@ -9,6 +43,106 @@ export default {
...
+          currentPage = 1
+        }
+      }
+
+      const hasNextPage = currentPage < pageCount
+      const hasPreviousPage = currentPage > 1
+
+      return {
+        adminRetrievals: camelcaseKeys(data, { deep: true }),
+        pageInfo: {
```
Comment (macrouch) [inline serverless/src/graphQl/resolvers/admin.js:68]: I'm trying to decide if I think `pageInfo` and `count` should be their own resolvers. If this were an open API I'd say definitely, so we wouldn't perform the logic involved unless the user wanted that data. But in our closed environment we will always be requesting those fields, so it doesn't hurt to be here. 

I wonder if in some of the future efforts we'll want to pull them out into a resolver, or potentially a util function?

File: serverless/src/graphQl/resolvers/admin.js [hunk 18]
```diff
@@ -9,6 +43,106 @@ export default {
...
+        },
+        count: retrievalCount
+      }
+    }
+  },
+  AdminRetrieval: {
+    user: async (parent, args, context) => {
+      const { loaders } = context
+
+      const loaderData = await loaders.users.load(parent.userId)
```
Comment (macrouch) [inline serverless/src/graphQl/resolvers/admin.js:84]: Because we haven't used dataloader very much, add a comment on what this is doing (and in `retrievalCollections` and `retrievalOrders`)

File: serverless/src/graphQl/utils/__tests__/databaseClient.test.js [hunk 19]
```diff
@@ -37,6 +37,718 @@ describe('DatabaseClient', () => {
...
+
+      const { queries } = dbTracker.queries
+
+      expect(queries[0].sql).toEqual('select "retrievals".* from "retrievals" where "retrievals"."id" = $1 limit $2')
+      expect(queries[0].bindings).toEqual([1, 1])
+      expect(queries[0].method).toEqual('first')
+    })
+
+    // TODO why does this console.log not being called the same way as the others
+    test.skip('returns an error', async () => {
```
Comment (macrouch) [inline serverless/src/graphQl/utils/__tests__/databaseClient.test.js]: Lets figure this out

File: serverless/src/graphQl/utils/__tests__/getLoaders.test.js [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+const mockContext = {
+  databaseClient: {
+    getRetrievalCollectionsByRetrievalId: jest.fn(),
+    getRetrievalOrdersByRetrievalCollectionId: jest.fn(),
+    getUsersById: jest.fn()
+  }
+}
+
+describe('getLoaders', () => {
+  it('returns the correct loaders', () => {
```
Comment (macrouch) [inline serverless/src/graphQl/utils/__tests__/getLoaders.test.js]: `test`, not `it`

File: static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx [hunk 21]
```diff
@@ -1,180 +1,157 @@
-import React from 'react'
 import {
-  render,
   screen,
+  waitFor,
   within
 } from '@testing-library/react'
 import userEvent from '@testing-library/user-event'
```
Comment (macrouch) [inline static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx]: `user` should come from `const { user } = setup()`, and shouldn't need to be imported here

File: serverless/src/graphQl/utils/__tests__/databaseClient.test.js (hunk 19 above)
Comment (macrouch) [reply serverless/src/graphQl/utils/__tests__/databaseClient.test.js]: Working test:
```
      const errorMessage = 'Failed to retrieve retrieval orders by ID'
      console.log(errorMessage)
```

This test:
```
      const errorMessage = 'Failed to retrieve retrieval by ID'
      console.log(e)
```

File: serverless/src/graphQl/resolvers/admin.js (hunk 17 above)
Comment (trevorlang) [reply serverless/src/graphQl/resolvers/admin.js:68]: Yeah, I think they will need to be in the next ticket when I am dealing with pagination again. It would be unfortunate to have that logic exist in each resolver.

File: serverless/src/graphQl/utils/__tests__/databaseClient.test.js (hunk 19 above)
Comment (trevorlang) [reply serverless/src/graphQl/utils/__tests__/databaseClient.test.js]: Ah, good catch... This was a sneaky one.

File: serverless/src/graphQl/utils/__tests__/getLoaders.test.js (hunk 20 above)
Comment (trevorlang) [reply serverless/src/graphQl/utils/__tests__/getLoaders.test.js]: Copilotttttt :shakesfist:

File: serverless/src/graphQl/resolvers/admin.js [hunk 22]
```diff
@@ -1,5 +1,39 @@
+import camelcaseKeys from 'camelcase-keys'
+import { obfuscateId } from '../../util/obfuscation/obfuscateId'
 import formatAdminPreferencesMetrics from '../utils/formatAdminPreferencesMetrics'
 
+// Const paginateData = async (query, key, { limit, offset }) => {
```
Comment (eudoroolivares2016) [inline serverless/src/graphQl/resolvers/admin.js]: Can we just remove this unless you want it for reference in the next one

File: static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx [hunk 23]
```diff
@@ -1,180 +1,154 @@
...
+      await waitFor(() => {
+        expect(screen.getByText('06347346')).toBeInTheDocument()
       })
 
-      renderContainer(props)
       await user.click(screen.getByRole('button', { name: /Requeue/ }))
 
       expect(props.onRequeueOrder).toHaveBeenCalledTimes(1)
-      expect(props.onRequeueOrder).toHaveBeenCalledWith(5)
+      expect(props.onRequeueOrder).toHaveBeenCalledWith(42)
```
Comment (eudoroolivares2016) [inline static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx:151]: I'm confused why this is called so many times here

File: static/src/js/actions/index.js [hunk 24]
```diff
@@ -1,13 +1,7 @@
...
-import {
-  adminViewRetrieval,
-  fetchAdminRetrievals,
-  fetchAdminRetrieval,
-  requeueOrder,
-  updateAdminRetrievalsSortKey,
-  updateAdminRetrievalsPageNum
-} from './admin/retrievals'
+
+import { requeueOrder } from './admin/retrievals'
```
Comment (eudoroolivares2016) [inline static/src/js/actions/index.js:4]: I'm having issues getting requeue to work locally i.e. the toast does not get brought up. I started edsc with all the optionals during the test so elasticmq is running it might be easier to test it in SIT to be sure but, on main the toast does come up

Comment (eudoroolivares2016) [reply static/src/js/actions/index.js:4]: image
Yea locally on main requeue and then refreshing I can see the order go back into previous state and reruns the job on Harmony or example
image

File: static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx (hunk 23 above)
Comment (trevorlang) [reply static/src/js/components/AdminRetrievalDetails/__tests__/AdminRetrievalDetails.test.jsx:151]: Its not called 42 times. Its called _with_ 42 which is the id of the mocked order 😂

File: static/src/js/actions/index.js (hunk 24 above)
Comment (trevorlang) [reply static/src/js/actions/index.js:4]: Fixed this issue in the latest commit ✅
